### PR TITLE
React SDK

### PR DESCRIPTION
**Pull request by divyamlamiyan**

### CODE REVIEW

File: example/react/custom-bar-chart/App.css
Comment (chetan1507) [inline example/react/custom-bar-chart/App.css:1]: Remove empty files

File: example/react/custom-bar-chart/assets/react.svg
Comment (chetan1507) [inline example/react/custom-bar-chart/assets/react.svg:1]: remove

File: example/react/custom-bar-chart/package.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "vite-project",
```
Comment (chetan1507) [inline example/react/custom-bar-chart/package.json:2]: update name.

File: example/react/custom-bar-chart/public/vite.svg
Comment (chetan1507) [inline example/react/custom-bar-chart/public/vite.svg:1]: remove

File: pnpm-lock.yaml
Comment (chetan1507) [inline pnpm-lock.yaml:1]: lets update package locks too.

File: src/react/custom-chart.context.tsx [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    extends ChartToTSEventEmitters,
+        TSToChartEventListener {
+    hasInitialized: boolean;
+    chartModel: ChartModel | undefined;
+}
+
+interface ChartProviderProps {
+    contextProps: CustomChartContextProps;
+    children: any;
+}
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:39]: Move all types and interface to a types.ts file inside react.

File: src/react/custom-chart.context.tsx [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+const emitter = (ctx: CustomChartContext) => {
+    function reducerFn(
+        acc: ChartToTSEventEmitters,
+        eventKey: string,
+    ): ChartToTSEventEmitters {
+        // TODO: do a better mapping for ChartToTSEvent and PayloadMap
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:46]: whats pending here? can we close this?

File: src/react/custom-chart.context.tsx [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    hasInitialized: boolean;
+    chartModel: ChartModel | undefined;
+}
+
+interface ChartProviderProps {
+    contextProps: CustomChartContextProps;
+    children: any;
+}
+
+const emitter = (ctx: CustomChartContext) => {
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:41]: return type is not defined.
Please add detailed jsdoc comments to each part of the code.

File: src/react/custom-chart.context.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    TSToChartEventsPayloadMap,
+} from '../types/ts-to-chart-event.types';
+
+export type ChartToTSEventEmitters = {
+    [key in keyof ChartToTSEventsPayloadMap as `emit${Capitalize<key>}`]: (
+        args: ChartToTSEventsPayloadMap[key],
+    ) => Promise<void>;
+};
+
+export type TSToChartEventListener = {
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:23]: Please add detailed jsdoc comments to each part of the code.

File: src/react/custom-chart.context.tsx [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        }
+        return acc;
+    }
+    return Object.keys(ChartToTSEvent).reduce(
+        reducerFn,
+        {} as ChartToTSEventEmitters,
+    );
+};
+
+const eventListener = (ctx: CustomChartContext) => {
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:66]: same as above

Also lets move emitter and eventListeners to different files each.

File: src/react/custom-chart.context.tsx [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        }
+        return acc;
+    }
+    return Object.keys(TSToChartEvent).reduce(
+        reducerFn,
+        {} as TSToChartEventListener,
+    );
+};
+
+const ChartContext = createContext<ChartContextProps>({} as ChartContextProps);
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:91]: ReactChartContext ?

File: src/react/custom-chart.context.tsx [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    ) => Promise<void>;
+};
+
+export type TSToChartEventListener = {
+    [key in keyof TSToChartEventsPayloadMap as `on${Capitalize<key>}`]: (
+        args: TSToChartEventsPayloadMap[key],
+    ) => Promise<void>;
+};
+
+interface ChartContextProps
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:29]: ReactChartContextProps

File: src/react/custom-chart.context.tsx [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    extends ChartToTSEventEmitters,
+        TSToChartEventListener {
+    hasInitialized: boolean;
+    chartModel: ChartModel | undefined;
+}
+
+interface ChartProviderProps {
+    contextProps: CustomChartContextProps;
+    children: any;
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:38]: this cannot be any

File: src/react/custom-chart.context.tsx [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+    const initializeProvider = async (ctx2: CustomChartContext) => {
+        ctx2.initialize()
+            .then(() => {
+                setHasInitialized(true);
+                setContextState(ctx2);
+                return true;
+            })
+            .catch((e) => {
+                console.log(e);
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:110]: can we add more meaningful error message?
Also is this an error? lets do console error if thats the case.
Should we not throw the error here?

File: src/react/custom-chart.context.tsx [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+const ChartProvider: React.FC<ChartProviderProps> = ({
+    contextProps,
+    children,
+}: ChartProviderProps) => {
+    const [hasInitialized, setHasInitialized] = useState(false);
+    const [ctx, setContextState] = useState<CustomChartContext>();
+    const [chartModel, setChartModel] = useState<ChartModel>();
+    const chartRef = useRef(null);
+
+    const initializeProvider = async (ctx2: CustomChartContext) => {
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:102]: please add comments for all the state variables and every function

File: src/react/custom-chart.context.tsx [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            setChartModel(context.getChartModel());
+        });
+        // update data
+        getChartContextValues(context).onDataUpdate(() => {
+            setChartModel(context.getChartModel());
+        });
+    }, []);
+
+    return (
+        <ChartContext.Provider value={getChartContextValues(ctx as any)}>
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:148]: should not have any.

**Review by chetan1507 — CHANGES_REQUESTED** [review]

File: src/react/custom-chart.context.tsx
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:1]: can we rename file to chart context provider?

File: src/react/custom-chart.context.tsx [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+        eventKey: string,
+    ): ChartToTSEventEmitters {
+        // TODO: do a better mapping for ChartToTSEvent and PayloadMap
+        if (eventKey in ChartToTSEvent) {
+            const eventName = eventKey as keyof ChartToTSEventsPayloadMap;
+            const emitterKey =
+                `emit${eventKey}` as keyof ChartToTSEventEmitters;
+
+            acc[emitterKey] = (
+                args: ChartToTSEventsPayloadMap[typeof eventName],
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:53]: there can be multiple args here right? That doesn't seem to have been taken care of.

File: src/react/custom-chart.context.tsx [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+        {} as ChartToTSEventEmitters,
+    );
+};
+
+const eventListener = (ctx: CustomChartContext) => {
+    function reducerFn(
+        acc: TSToChartEventListener,
+        eventKey: string,
+    ): TSToChartEventListener {
+        // TODO: do a better mapping for ChartToTSEvent and PayloadMap
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:71]: can we resolve this?

File: src/react/custom-chart.context.tsx [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+        if (eventKey in TSToChartEvent) {
+            const eventName = eventKey as keyof TSToChartEventsPayloadMap;
+            const emitterKey = `on${eventKey}` as keyof TSToChartEventListener;
+
+            acc[emitterKey] = (
+                callbackFn: TSToChartEventsPayloadMap[typeof eventName],
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:77]: this doesnt look like a callback function and looks more like object map. is this accurate?

File: src/react/custom-chart.context.tsx [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+const ChartContext = createContext<ChartContextProps>({} as ChartContextProps);
+
+const ChartProvider: React.FC<ChartProviderProps> = ({
+    contextProps,
+    children,
+}: ChartProviderProps) => {
+    const [hasInitialized, setHasInitialized] = useState(false);
+    const [ctx, setContextState] = useState<CustomChartContext>();
+    const [chartModel, setChartModel] = useState<ChartModel>();
+    const chartRef = useRef(null);
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:100]: remove

File: src/react/custom-chart.context.tsx (hunk 11 above)
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:102]: better name than ctx2?

File: src/react/custom-chart.context.tsx [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+        chartModel,
+        ...emitter(ctx),
+        ...eventListener(ctx),
+    });
+
+    React.useEffect(() => {
+        const context = new CustomChartContext({
+            ...contextProps,
+            renderChart: (ctx: CustomChartContext) => {
+                chartRef.current = children;
```
Comment (chetan1507) [inline src/react/custom-chart.context.tsx:127]: why do we need this?

Comment (chetan1507) [reply src/react/custom-chart.context.tsx:127]: how do we trigger render chart from backend and then it updates the same?

**Review by chetan1507 — CHANGES_REQUESTED** [review]

File: src/react/custom-chart-context-hook.tsx [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+                ...emitter(ctx),
+                ...eventListener(ctx),
+                WrapperComponent: ({ children }: WrapperComponentProps) => {
+                    return (
+                        <React.Fragment key={key}>{children}</React.Fragment>
+                    );
+                },
+            };
+        },
+        [chartModel, key, hasInitialized],
```
Comment (chetan1507) [inline src/react/custom-chart-context-hook.tsx:203]: ctx also.

File: src/react/custom-chart-context-hook.tsx [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+
+/**
+ * A custom hook to manage the Chart Context state and provide necessary
+ * chart-related functionality.
+ * @param {CustomChartContextProps} props - The custom chart context properties to initialize the chart context.
+ * @returns {ChartContextProps} The chart context values, including initialized state,
+ * chart model, emitter, event listener, and WrapperComponent.
+ */
+export const useChartContext = (
+    props: CustomChartContextProps,
```
Comment (chetan1507) [inline src/react/custom-chart-context-hook.tsx:139]: remove renderChart from here.

File: example/react/custom-bar-chart/App.tsx [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+                        key: 'datalabels',
+                        type: 'toggle',
+                        defaultValue: false,
+                        label: 'Data Labels',
+                    },
+                ],
+            },
+        ],
+    },
+    renderChart: (cx) => {
```
Comment (chetan1507) [inline example/react/custom-bar-chart/App.tsx:121]: remove this from renderChart from the render.

File: example/react/custom-bar-chart/line-chart.component.tsx
Comment (chetan1507) [inline example/react/custom-bar-chart/line-chart.component.tsx:1]: move this inside the src folder.

Comment (chetan1507) [reply example/react/custom-bar-chart/line-chart.component.tsx:1]: add comments as much as possible to help developers understand the example better.

File: example/react/custom-bar-chart/line-chart.component.tsx [hunk 21]
```diff
@@ -0,0 +1,281 @@
...
+        const columnChartModel = getColumnDataModel(
+            chartModel.config?.chartConfig?.[0].dimensions ?? [],
+            chartModel.data?.[0].data as DataPointsArray,
+            'line',
+            chartModel?.visualProps,
+        );
+
+        return columnChartModel;
+    }, [chartModel.config, chartModel.data, chartModel.visualProps]);
+    // useEffect(() => {
```
Comment (chetan1507) [inline example/react/custom-bar-chart/line-chart.component.tsx:183]: have we tested this?

File: example/react/custom-bar-chart/line-chart.component.tsx [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+        if (hasInitialized) {
+            emitRenderComplete([]);
+        }
+    }, [hasInitialized]);
+
+    if (!hasInitialized || !chartModel) {
+        return <div>Loading...</div>;
+    }
+
+    emitRenderStart([]);
```
Comment (chetan1507) [inline example/react/custom-bar-chart/line-chart.component.tsx:271]: only if intialized

**Review by chetan1507 — CHANGES_REQUESTED** [review]

File: example/react/custom-bar-chart/line-chart.component.tsx
Comment (divyamlamiyan) [reply example/react/custom-bar-chart/line-chart.component.tsx:1]: Done

File: example/react/custom-bar-chart/src/App.tsx [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+ * @function
+ * @name App
+ * @returns {React.FC} - The main application functional component.
+ */
+const App: React.FC = () => {
+    const ref = useRef(null);
+
+    const {
+        chartModel,
+        WrapperComponent,
```
Comment (ashubham) [inline example/react/custom-bar-chart/src/App.tsx:134]: TSChartContext

File: example/react/custom-bar-chart/src/App.tsx [hunk 24]
```diff
@@ -0,0 +1,169 @@
...
+        hasInitialized,
+        emitOpenContextMenu,
+        emitRenderStart,
+        emitRenderError,
+        emitRenderComplete,
+        onVisualPropsUpdate,
```
Comment (ashubham) [inline example/react/custom-bar-chart/src/App.tsx:140]: setOnVisualPropsUpdate

File: example/react/custom-bar-chart/src/App.tsx [hunk 25]
```diff
@@ -0,0 +1,169 @@
...
+        offVisualPropsUpdate,
```
Comment (ashubham) [inline example/react/custom-bar-chart/src/App.tsx:141]: setOffVisualPropsUpdate

File: example/react/custom-bar-chart/src/line-chart.component.tsx [hunk 26]
```diff
@@ -0,0 +1,299 @@
...
+    emitRenderStart,
+    emitRenderComplete,
+    emitOpenContextMenu,
+    offVisualPropsUpdate,
+    onVisualPropsUpdate,
+}: RenderChartProps) => {
+    useEffect(() => {
+        if (hasInitialized) {
+            emitRenderStart();
+            emitRenderComplete();
```
Comment (ashubham) [inline example/react/custom-bar-chart/src/line-chart.component.tsx:281]: `emitRenderComplete` should be called after the rendering of the chart is complete

File: example/react/custom-bar-chart/src/App.css
Comment (chetan1507) [inline example/react/custom-bar-chart/src/App.css:1]: Remove this file

File: example/react/custom-bar-chart/src/line-chart.component.tsx [hunk 27]
```diff
@@ -0,0 +1,300 @@
...
+    chartModel,
+    emitRenderStart,
+    emitRenderComplete,
+    emitOpenContextMenu,
+    setOffVisualPropsUpdate,
+    setOnVisualPropsUpdate,
+}: RenderChartProps) => {
+    useEffect(() => {
+        if (hasInitialized) {
+            emitRenderStart();
```
Comment (chetan1507) [inline example/react/custom-bar-chart/src/line-chart.component.tsx:276]: https://www.chartjs.org/docs/3.9.1/api/interfaces/Plugin.html#afterrender
Lets add a todo to leverage this.

File: src/react/use-custom-chart-context.spec.tsx [hunk 28]
```diff
@@ -0,0 +1,500 @@
...
+        // Check if the child element is still in the document after re-render
+        const updatedChildElement = getByTestId('child-element');
+        await waitFor(() => {
+            expect(updatedChildElement.textContent).toBe(
+                'Child Element counter: 2red',
+            );
+        });
+    });
+
+    // test.only('TSChartContext renders children should increase counter for
```
Comment (chetan1507) [inline src/react/use-custom-chart-context.spec.tsx:466]: can we fix this?

File: src/react/use-custom-chart-context.tsx [hunk 29]
```diff
@@ -0,0 +1,251 @@
...
+        return acc;
+    }, {} as TSToChartEventListener);
+};
+
+/**
+ *
+ * @param ctx Custom Chart Context
+ * @returns List of event listeners
+ */
+const evenOfftListener = (
```
Comment (chetan1507) [inline src/react/use-custom-chart-context.tsx:95]: spelling?

File: src/react/use-custom-chart-context.tsx [hunk 30]
```diff
@@ -0,0 +1,251 @@
...
+    TSToChartEventListener,
+    TSToChartEventOffListener,
+} from './use-custom-chart-types';
+
+/**
+ *
+ * @param ctx Custom Chart Context
+ * @returns List of event emitters
+ */
+const emitter = (ctx: CustomChartContext): ChartToTSEventEmitters => {
```
Comment (chetan1507) [inline src/react/use-custom-chart-context.tsx:40]: move to utils

File: src/react/use-custom-chart-context.tsx [hunk 31]
```diff
@@ -0,0 +1,251 @@
...
+        return acc;
+    }, {} as ChartToTSEventEmitters);
+};
+
+/**
+ *
+ * @param ctx Custom Chart Context
+ * @returns List of event listeners
+ */
+const eventListener = (ctx: CustomChartContext): TSToChartEventListener => {
```
Comment (chetan1507) [inline src/react/use-custom-chart-context.tsx:67]: move to utils

File: src/react/use-custom-chart-context.tsx (hunk 29 above)
Comment (chetan1507) [reply src/react/use-custom-chart-context.tsx:95]: move to utils

File: src/react/use-custom-chart-context.tsx [hunk 32]
```diff
@@ -0,0 +1,251 @@
...
+    useEffect(() => {
+        // Create a new custom chart context with the provided props and
+        // renderChart function.
+        const context = new CustomChartContext({
+            ...props,
+            renderChart,
+        });
+        // Initialize the chart context provider.
+        initializeContext(context);
+        setupEventListeners(context);
```
Comment (chetan1507) [inline src/react/use-custom-chart-context.tsx:247]: we ll also need to call off function here too right? to destroy the chart?

**Review by chetan1507 — CHANGES_REQUESTED** [review]

**Pull request merged**